bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2460,7 +2460,7 @@ void BCStateTran::processData(bool lastInBatch) {
   bool badDataFromCurrentSourceReplica = false;
 
   while (true) {
-    bool newSourceReplica = sourceSelector_.shouldReplaceSource(currTime, badDataFromCurrentSourceReplica);
+    bool newSourceReplica = sourceSelector_.shouldReplaceSource(currTime, badDataFromCurrentSourceReplica, lastInBatch);
```

Comment:
You may end up choosing new source just before processing last chunk from batch.